packages/cursorless-engine/src/processTargets/modifiers/surroundingPair/index.ts

```diff
@@ -140,14 +102,41 @@ function processSurroundingPairCore(
     }
   }
 
-  // If we have a parse tree but we are in a string node or in a comment node,
-  // then we use the text-based algorithm
-  const selectionWithEditor = {
-    editor,
-    selection: new Selection(range.start, range.end),
-  };
-  const textFragmentRange = textFragmentExtractor(node, selectionWithEditor);
+  const textFragmentRange = (() => {
+    // First try to use the text fragment scope handler if it exists
+    const textFragmentScopeHandler =
+      languageDefinition?.getTextFragmentScopeHandler();
+
+    if (textFragmentScopeHandler != null) {
+      const containingScope = getContainingScopeTarget(
+        target,
+        textFragmentScopeHandler,
+        0,
+      );
+
+      return containingScope?.[0].contentRange;
+    }
+
+    // Then try to use the legacy text fragment extractor if it exists
+    const textFragmentExtractor = getTextFragmentExtractor(document.languageId);
+
+    if (textFragmentExtractor == null) {
+      // If the text fragment extractor doesn't exist, or if it explicitly is
+      // set to `null`, then we just use text-based algorithm on entire document
+      return document.range;
+    }
+
+    const selectionWithEditor = {
+      editor,
+      selection: new Selection(range.start, range.end),
```

Comment:
i think we have a range.toSelection. although looking at it, it has a boolean arg, ick. (maybe we could make that optional/named and default to false?)

Reply:
I think it's useful to force the user to think about the direction. If you don't care about a direction a selection is not that different from a range.

Reply:
sure. but at the call site `.toSelection(false)` is inscrutable.